staging/src/k8s.io/apiserver/pkg/server/filters/timeout.go

```diff
@@ -57,14 +58,18 @@ func WithTimeoutForNonLongRunningRequests(handler http.Handler, longRunning apir
 
 		postTimeoutFn := func() {
 			cancel()
-			metrics.Record(req, requestInfo, "", http.StatusGatewayTimeout, 0, 0)
+			contentType := w.Header().Get("Content-Type")
```

Comment:
as far as I see this is not race free. The headers are owned by the handler's go routine. Though the chance for a race is probably very low.

Reply:
The question is what can happen as a result of that race - I think we're risking setting JSON where in fact in the meantime we set Content-Type. I think that is fine.

Or do you think that this part should be reverted?

Reply:
"Accept" headers won't help you, will they? You will see json and proto in them I guess.

Reply:
Also: our List implementation in the handler https://github.com/kubernetes/kubernetes/blob/862f8567bb4b1d4e9f63330cf3a516af61ae7cc5/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/get.go#L167 is not using streaming of any kind. How much of the timeout (60 sec) is the actual encoding (which depends on the content type)?

Reply:
> "Accept" headers won't help you, will they? You will see json and proto in them I guess.

Our client libraries by default sets "your-preffered-content-type,*/*" (but you're allowed to overwrite it). So yeah - we may use that, but that's not great (though better than empty).

Actually, thinking more about it, given that it isn't racy, maybe it's a better choice. WDYT?

> How much of the timeout (60 sec) is the actual encoding (which depends on the content type)?

It really depends (on amount of data, content-type, available cpu, etc.)

Reply:
`your-preffered-content-type` is proto probably even for CRDs which don't support it. So it tends to be misleading.

Thinking out loud:

If the handler is still in the reading-from-etcd phase, `w.Header` will be empty. Only if we have started writing anything, the header is set. And then we also have no race anymore because the http header was sent and the body is being written.

You could wrap the Header() func of the response writer and intercept the content-type. This would be race-free if you use atomic or a lock. How about that?

Reply:
Another thought: you could also return a more precise error, depending on which phase the handler is in, e.g. reading+processing or sending of the body.

Reply:
> You could wrap the Header() func of the response writer and intercept the content-type. This would be race-free if you use atomic or a lock. How about that?

A bit questionable as well: `Header()` gives directly a map, with accessor methods though.